### ISSUE TITLE
DST and getNextRunDate skips day

### ISSUE THREAD

**Issue by einarsozols** [dragonmantank/cron-expression#154]
`America/New_York` time zone changed to daylight saving time at 2:00 AM on Sunday, Mar 12, 2023.

```php
// php version 8.1.16, package version 3.3.2
$date = new DateTime('2023-03-11 08:00:00', new DateTimeZone('America/New_York'));
$cron = new CronExpression('30 07 * * *');
$nextDate = $cron->getNextRunDate($date);
```

Expected date: `2023-03-12 07:30:00.0 America/New_York (-04:00)`
Returned date: `2023-03-13 07:30:00.0 America/New_York (-04:00)`

It works fine for other days, for example, if check would happen every day at 08:00 starting from 8th march:

```
2023-03-09 07:30:00.0 America/New_York (-05:00)
2023-03-10 07:30:00.0 America/New_York (-05:00)
2023-03-11 07:30:00.0 America/New_York (-05:00)
// Missing 12th march
2023-03-13 07:30:00.0 America/New_York (-04:00)
2023-03-14 07:30:00.0 America/New_York (-04:00)
2023-03-15 07:30:00.0 America/New_York (-04:00)
```

**Comment by holtkamp** [comment]
Might be related to https://github.com/dragonmantank/cron-expression/issues/133 ?

**Comment by einarsozols** [comment]
I tested package `3.3.2` with php 8.1, 7.4, 7.2
All of them have the same issue as I described in my issue.

Tried to test php 8.1 with different package versions.

`3.3.2`: FAIL
`3.3.1`: FAIL
`3.3.0`: FAIL
`3.2.4`: FAIL
`3.2.3`: FAIL
`3.2.2`: FAIL
`3.2.1`: FAIL
`3.2.0`: **SUCCESS**

So I guess the problem is somewhere here: https://github.com/dragonmantank/cron-expression/compare/v3.2.0...dragonmantank:cron-expression:v3.2.1

**Comment by holtkamp** [comment]
Indeed, I suspect this function https://github.com/dragonmantank/cron-expression/blob/782ca5968ab8b954773518e9e49a6f892a34b2a8/src/Cron/HoursField.php#L123-L129

**Comment by Dr-Shadow** [comment]
@dragonmantank Would be it possible to get a fix for the next daylight saving time schedule which would be around October/November ?

**Comment by dragonmantank** [comment]
3.2.1 did introduce a change to DST handling, as there were instances were it was skipping (particularly in +1 timezones and the London timezone). There was also a change in how DST is handled in PHP ~8.0 sometime, which broke how it was handled in prior versions.

I'll see what I can do since this is coming up, but part of the fix _might_ be just dropping older PHP support to simplify DST handling.

**Comment by curunoir** [comment]
Hello, is this issue resolved in 3.3.3 ?

**Comment by daum** [comment]
Don't believe so we just hit it.

**Comment by plsoucy-tapclicks** [comment]
If we are only using this in North American time zones, would downgrading to 3.2.0 be a good way to prevent the issue until you find a long-term fix? Just looking for options to make sure we don't get impacted on the next DST change.

Thanks for your help!

**Comment by daum** [comment]
We downgraded to that and I tested out a few different scenarios using timecop and it worked for us!

**Comment by dragonmantank** [comment]
I believe I have a fix for this, it is due to how time math was being handled. Before I release a fix, could I get your thoughts on https://github.com/dragonmantank/cron-expression/discussions/181 as it will impact the overall solution?

**Comment by sc0ttdav3y** [comment]
We had this happen last October for `Australia/Sydney`, and we're expecting it to happen again this weekend. We're on 3.3.3. I only discovered this issue today. We'll try out 3.2.0 as a workaround for now. 

@dragonmantank, thank you for this library and your work in maintaining it ❤️. If you had to drop 7.4 to simplify the internals, I'd be fine with that, and I say this as an engineer that still maintains some legacy 7.4 systems.
